image/helpers.go

```diff
@@ -206,38 +208,81 @@ func NewToolingStore() (*ToolingStore, error) {
 
 // DownloadOptions carries options for downloading snaps plus assertions.
 type DownloadOptions struct {
+	Revision  snap.Revision
 	TargetDir string
 	Channel   string
+	CohortKey string
 }
 
-// DownloadSnap downloads the snap with the given name and optionally revision  using the provided store and options. It returns the final full path of the snap inside the opts.TargetDir and a snap.Info for the snap.
-func (tsto *ToolingStore) DownloadSnap(name string, revision snap.Revision, opts *DownloadOptions) (targetFn string, info *snap.Info, err error) {
-	if opts == nil {
-		opts = &DownloadOptions{}
+var errRevisionAndCohort = errors.New("cannot specify both revision and cohort")
+
+func (opts *DownloadOptions) validate() error {
+	if opts.Revision.Unset() || opts.CohortKey == "" {
+		return nil
+	}
+	return errRevisionAndCohort
+}
+
+// TODO: maybe move this to strutil next to ElliptRight
+func elliptLeft(str string) string {
+	if len(str) < 10 {
+		// shouldn't happen outside of tests
+		return str
+	}
+	return "…" + str[len(str)-8:]
+}
+
+func (opts *DownloadOptions) String() string {
+	var spec []string
+	if !opts.Revision.Unset() {
+		spec = append(spec, fmt.Sprintf("(%s)", opts.Revision))
+	}
+	if opts.Channel != "" {
+		spec = append(spec, fmt.Sprintf("from channel %q", opts.Channel))
+	}
+	if opts.CohortKey != "" {
+		// cohort keys are really long, and the rightmost bit being the
+		// interesting bit, so ellipt the rest
+		spec = append(spec, fmt.Sprintf(`from cohort %q`, elliptLeft(opts.CohortKey)))
+	}
+	if opts.TargetDir != "" {
+		spec = append(spec, fmt.Sprintf("to %q", opts.TargetDir))
+	}
+	return strings.Join(spec, " ")
+}
+
+// DownloadSnap downloads the snap with the given name and optionally revision
+// using the provided store and options. It returns the final full path of the
+// snap inside the opts.TargetDir and a snap.Info for the snap.
+func (tsto *ToolingStore) DownloadSnap(name string, opts DownloadOptions) (targetFn string, info *snap.Info, err error) {
+	if err := opts.validate(); err != nil {
+		return "", nil, err
 	}
 	sto := tsto.sto
 
-	targetDir := opts.TargetDir
-	if targetDir == "" {
+	if opts.TargetDir == "" {
 		pwd, err := os.Getwd()
 		if err != nil {
 			return "", nil, err
 		}
-		targetDir = pwd
+		opts.TargetDir = pwd
 	}
 
-	logger.Debugf("Going to download snap %q (%s) from channel %q to %q.", name, revision, opts.Channel, opts.TargetDir)
+	if !opts.Revision.Unset() {
+		// XXX: is this really necessary (and, if it is, shoudn't we error out instead)
+		opts.Channel = ""
```

Comment:
it's a long story, there are aspect to the handling of this in various places, first:

* snap install --channel=foo --revision=9999 in theory is ok  it gets a revision but is also sets the tracking channel, do we want to let people do that with download
* not relevant to this code here, but in general we had a bug at some point by which we ended up with snapst.TrackingChannel == ""